src/utils/printing/getPrinterIppAttributes.ts

```diff
@@ -0,0 +1,261 @@
+  'marker-supply-empty': 'The printer is out of ink.',
+  'marker-waste-almost-full':
+    "The printer's ink waste receptacle is almost full.",
+  'marker-waste-full': "The printer's ink waste receptacle is full.",
+  'fuser-over-temp': "The printer's fuser temperature is above normal.",
+  'fuser-under-temp': "The printer's fuser temperature is below normal.",
+  'opc-near-eol': "The printer's optical photo conductor is near end of life.",
+  'opc-life-over':
+    "The printer's optical photo conductor is no longer functioning.",
+  'developer-low': 'The printer is low on developer.',
+  'developer-empty': 'The printer is out of developer.',
+  'interpreter-resource-unavailable': 'An interpreter resource is unavailable.',
+}
+// IppPrinterStateReason only covers the basic RFC spec reasons - there may be
+// others, so we can't type them all.
+export type IppPrinterStateReason =
+  | keyof typeof IppPrinterStateReasonMessage
+  | string
+
+/**
+ * "Marker" is a general name for ink/toner/etc. CUPS implements a variety of
+ * marker-related IPP attributes prefixed with "marker-", e.g. "marker-levels".
+ * Each attribute is a comma-delimated list of values, since a printer may have
+ * multiple marker supplies (e.g. black and color ink cartridges). Here, we
+ * represent the marker info for a single marker supply.
+ * Spec: https://www.cups.org/doc/spec-ipp.html
+ */
+export interface IppMarkerInfo {
+  name: string // e.g. "black cartridge"
+  color: string // e.g. "#000000"
+  type: string // e.g. "toner-cartridge"
+  lowLevel: number // e.g. 2
+  highLevel: number // e.g. 100
+  level: number // e.g. 83
+}
+
+/**
+ * A collection of status attributes we can get from a printer via IPP.
+ */
+export interface PrinterIppAttributes {
+  state?: IppPrinterState
+  stateReasons: IppPrinterStateReason[]
+  markerInfos: IppMarkerInfo[]
+}
+
+const ippAttributesToQuery = [
+  'printer-state',
+  'printer-state-reasons',
+  'marker-names',
+  'marker-colors',
+  'marker-types',
+  'marker-low-levels',
+  'marker-high-levels',
+  'marker-levels',
+]
+const ippQuery = `{
+  OPERATION Get-Printer-Attributes
+  GROUP operation-attributes-tag
+  ATTR charset attributes-charset utf-8
+  ATTR language attributes-natural-language en
+  ATTR uri printer-uri $uri
+  ATTR keyword requested-attributes ${ippAttributesToQuery.join(',')}
+}`
+
+/**
+ * Query the printer for its status via IPP.
+ */
+export async function getPrinterIppAttributes(
+  printerIppUri: string,
+): Promise<PrinterIppAttributes> {
+  // Tricky business: ipptool takes a query in file form, so we feed the query
+  // string to stdin, and then pass /dev/stdin as the "file". Since node child
+  // process file descriptors are sockets, not pipes, we have to use `sh` to run
+  // the command in order to get /dev/stdin to work.
+  // See https://github.com/nodejs/node-v0.x-archive/issues/3530#issuecomment-6561239
+  const ipptoolCommand = `ipptool -tv ${printerIppUri} /dev/stdin`
```

Comment:
For security reasons, I'd prefer creating a temporary file over interpolating the `printerIppUri` value into the command. However, if there's a way to use `/dev/stdin` while also delegating the command-injection protection to `child_process`, I'm up for it.

Reply:
Are you suggesting creating a temporary file for the query so that we don't have to pass it to `stdin`? We could even just commit the query file to the repo, though I'm not sure if it would end up in the right place during build if that gets weird.

Reply:
I forgot that it really could just be a static file. I think we may be able to rely on the build to copy files over. Maybe try that out?